File: bot/src/service/UsagePlanService.ts
import {Context} from "../Context";
import UsagePlanOrmService from "../orm/UsagePlanOrmService";
import ServiceError from "../ServiceError";

export type UsagePlan = {
    id: string,
    title: string,
    maxCredits: number,
    creditGainAmount: number,
    creditGainPeriodSeconds: number,
    visible: boolean
};

const ROUTINE_INTERVAL = 15 * 1000; // 15 seconds
const DEFAULT_USAGE_PLAN_ID = "default";

export default class UsagePlanService {
    private orm!: UsagePlanOrmService;

    private intervalId: NodeJS.Timeout | null = null;

    constructor(private context: Context) {
        context.onReady(() => {
            this.orm = this.context.usagePlanOrmService;
            this.intervalId = setInterval(this.doRoutineTasks.bind(this), ROUTINE_INTERVAL);
        });
    }

    async getDefaultUsagePlan(): Promise<UsagePlan | null> {
        return this.orm.getUsagePlan(DEFAULT_USAGE_PLAN_ID);
    }

    async getUsagePlanOrDefault(id: string | null): Promise<UsagePlan | null> {
        if (id == null)
            return this.getDefaultUsagePlan();
        return await this.orm.getUsagePlan(id);
    }

    async doRoutineTasks() {
        try {
            await this.checkPlansExpiration();
        } catch (e) {
            console.error('Failed to do routine tasks:', e);
        }
        try {
            await this.gainCreditsRoutine();
        } catch (e) {
            console.error('Failed to do routine tasks:', e);
        }
    }

    async checkPlansExpiration() {
        await this.context.vkUsersOrmService.setUserUsagePlanWhereExpiryIsBefore(null, new Date());
    }

    async gainCreditsRoutine() {
        const now = new Date();
        const users = await this.context.vkUsersOrmService.getUsersWithCreditsLessThanMax();
        const plans = new Map((await this.orm.getUsagePlans()).map(plan => [plan.id, plan] as [string, UsagePlan]));
        for (const user of users) {
            const planId = user.usagePlanId || DEFAULT_USAGE_PLAN_ID;
            if (planId == null || !plans.has(planId))
                continue;
            const plan = plans.get(planId)!;
            if (user.credits >= plan.maxCredits || plan.creditGainPeriodSeconds <= 0)
                continue;
            const secondsSinceLastCreditGain = Math.floor((now.getTime() - user.lastCreditGain.getTime()) / 1000);
            const gainCreditTimes = Math.floor(secondsSinceLastCreditGain / plan.creditGainPeriodSeconds);
            const creditsToGain = gainCreditTimes * plan.creditGainAmount;
            user.credits += creditsToGain;
            if (user.credits > plan.maxCredits)
                user.credits = plan.maxCredits;
            user.lastCreditGain = new Date(
                user.lastCreditGain.getTime() + gainCreditTimes * plan.creditGainPeriodSeconds * 1000
            );
            await this.context.vkUsersOrmService.saveUser(user);
        }
        await this.context.vkUsersOrmService.refreshLastCreditGainWhenMaxCredits(now);
    }

    async getTimeInSecondsRequiredToHaveCredits(userId: number, creditsAmount: number): Promise<number> {
        const user = await this.context.vkUsersService.getUser(userId);
        if (!user)
            throw new ServiceError("User not found");
        if (user.credits >= creditsAmount)
            return 0;
        const plan = await this.orm.getUsagePlan(user.usagePlanId || DEFAULT_USAGE_PLAN_ID);
        if (!plan || plan.creditGainPeriodSeconds <= 0 || plan.creditGainAmount <= 0)
            return Number.POSITIVE_INFINITY;
        const creditsNeeded = creditsAmount - user.credits;
        const fullCyclesNeeded = Math.ceil(creditsNeeded / plan.creditGainAmount);
        const secondsSinceLastCreditGain = Math.floor((Date.now() - user.lastCreditGain.getTime()) / 1000);
        return fullCyclesNeeded * plan.creditGainPeriodSeconds - secondsSinceLastCreditGain;
    }
}

File: backend/src/main/java/space/davids_digital/sweetie/service/UsagePlanService.kt
package space.davids_digital.sweetie.service

import jakarta.transaction.Transactional
import org.slf4j.LoggerFactory
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service
import space.davids_digital.sweetie.model.UsagePlanModel
import space.davids_digital.sweetie.orm.repository.VkUserRepository
import space.davids_digital.sweetie.orm.service.UsagePlanOrmService
import space.davids_digital.sweetie.orm.service.VkUserOrmService
import java.time.ZonedDateTime
import kotlin.math.ceil
import kotlin.math.floor

@Service
class UsagePlanService(
    private val usagePlanOrmService: UsagePlanOrmService,
    private val vkUserRepository: VkUserRepository,
    private val vkUserOrmService: VkUserOrmService
) {
    companion object {
        private val log = LoggerFactory.getLogger(UsagePlanService::class.java)
        private const val DEFAULT_PLAN_ID = "default"
        private const val ROUTINE_DELAY = 15_000L
    }

    fun getDefault(): UsagePlanModel {
        val plan = usagePlanOrmService.getById(DEFAULT_PLAN_ID)
        if (plan == null) {
            log.warn("Default usage plan with id = '$DEFAULT_PLAN_ID' not found, falling back to internal default")
            return UsagePlanModel(
                id = DEFAULT_PLAN_ID,
                title = "Default",
                maxCredits = 0,
                creditGainAmount = 0,
                creditGainPeriodSeconds = 0,
                visible = false
            )
        }
        return plan
    }

    fun getOrDefault(planId: String?): UsagePlanModel {
        return usagePlanOrmService.getById(planId ?: DEFAULT_PLAN_ID) ?: getDefault()
    }

    @Scheduled(fixedDelay = ROUTINE_DELAY)
    fun checkPlansExpirationRoutine() {
        log.debug("Checking usage plans expiration")
        val expiredPlans = vkUserRepository.setUsagePlanWhereExpiryBefore(null, ZonedDateTime.now())
        if (expiredPlans > 0) {
            log.info("Removed $expiredPlans expired plans")
        }
    }

    @Transactional
    @Scheduled(fixedDelay = ROUTINE_DELAY)
    fun gainCreditsRoutine() {
        log.debug("Updating user credits")
        val now = ZonedDateTime.now()
        val users = vkUserOrmService.findAllByCreditsLessThanMax()
        val plans = usagePlanOrmService.findAll().associateBy { it.id }
        users.forEach { user ->
            val plan = plans[user.usagePlanId ?: DEFAULT_PLAN_ID] ?: return@forEach
            if (user.credits > plan.maxCredits || plan.creditGainPeriodSeconds <= 0) {
                return@forEach
            }
            val lastCreditGain = user.lastCreditGain?.toEpochSecond() ?: 0
            val secondsSinceLastCreditGain = floor((now.toEpochSecond() - lastCreditGain).toDouble())
            val gainCreditTimes = floor(secondsSinceLastCreditGain / plan.creditGainPeriodSeconds.toDouble())
            val creditsToGain = (gainCreditTimes * plan.creditGainAmount).toLong()
            var newCredits = user.credits + creditsToGain
            if (newCredits > plan.maxCredits)
                newCredits = plan.maxCredits
            val newLastCreditGain = user.lastCreditGain?.plusSeconds(
                (gainCreditTimes * plan.creditGainPeriodSeconds).toLong()
            ) ?: now
            vkUserOrmService.setCredits(user.id, newCredits)
            vkUserOrmService.setLastCreditGain(user.id, newLastCreditGain)
        }
        vkUserRepository.setLastCreditGainWhereMaxCredits(now)
    }

    fun getTimeInSecondsRequiredToHaveEnoughCredits(userId: Long, creditsRequired: Long): Long {
        val user = vkUserOrmService.getById(userId)
        if (user == null) {
            log.error("User with id $userId not found")
            return Long.MAX_VALUE
        }
        if (user.credits >= creditsRequired) {
            return 0
        }
        val plan = getOrDefault(user.usagePlanId)
        if (plan.creditGainAmount <= 0 || plan.creditGainPeriodSeconds <= 0) {
            return Long.MAX_VALUE
        }
        val creditsNeeded = creditsRequired - user.credits
        val fullCyclesNeeded = ceil(creditsNeeded.toDouble() / plan.creditGainAmount)
        val secondsSinceLastCreditGain = floor(
            ZonedDateTime.now().toEpochSecond().toDouble() - (user.lastCreditGain?.toEpochSecond() ?: 0)
        )
        return ceil(fullCyclesNeeded * plan.creditGainPeriodSeconds - secondsSinceLastCreditGain).toLong()
    }
}